### PR TITLE
Raise TypeError if component custom_id type is not `str`

### LINKED ISSUE #1010: Providing a custom id that isn't a string silently errors
### Summary

Instead of erroring verbosely the callback method never runs and discord interaction attempt times out

### Reproduction Steps

Initiate a button class (I presume the same issue exists with anything else that uses custom_id) with a string and define a custom_id of a type that is not string, Eg. int.

### Minimal Reproducible Code

```python
class TestButton(discord.ui.Button):
    def __init__(self):
        super().__init__(
            label="Test",
            style=discord.enums.ButtonStyle.primary,
            custom_id=1,
        )

    async def callback(self, interaction: discord.Interaction):
        print("Test")
        self.stop()
```

### Expected Results

Either there to be an error or for the id to be converted to a string, or the callback function to still run ignoring the invalid custom_id

### Actual Results

The callback function was never called and the discord interaction silently failed so I was left scratching my head

### Intents

Guilds, Messages, Emojis, Guild Reactions

### System Information

- Python v3.9.2-final
- py-cord v2.0.0-beta
    - py-cord pkg_resources: v2.0.0b4
- aiohttp v3.7.4.post0
- system info: Windows 10 10.0.22000 <- actually windows 11

### Checklist

- [X] I have searched the open issues for duplicates.
- [X] I have shown the entire traceback, if possible.
- [X] I have removed my token from display, if visible.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by krittick**
## Summary

This change raises a TypeError if a user provides a `custom_id` for a component (Button, Select, InputText, or Modal) that is not `str`. 

While `custom_id` is already type hinted to `str`, this change enforces it as a requirement instead of the current behavior where the interaction silently fails.

Closes #1010 

## Checklist

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why
- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...)

### CODE REVIEW

**Comment by Dorukyum** [comment]
`isinstance(x, SomeGenericType[y, z])` is not backwards compatible with python 3.8.

File: discord/ui/button.py [hunk 1]
```diff
@@ -99,6 +99,9 @@ def __init__(
         if custom_id is not None and url is not None:
             raise TypeError("cannot mix both url and custom_id with Button")
 
+        if not isinstance(custom_id, Union[str, None]):
```
Comment (Dorukyum) [inline discord/ui/button.py:102]:
```suggestion
        if not (isinstance(custom_id, str) or custom_id is None):
```

File: discord/ui/input_text.py [hunk 2]
```diff
@@ -58,6 +58,10 @@ def __init__(
     ):
         super().__init__()
         custom_id = os.urandom(16).hex() if custom_id is MISSING else custom_id
+
+        if not isinstance(custom_id, Union[str, None]):
```
Comment (Dorukyum) [inline discord/ui/input_text.py:62]:
```suggestion
        if not (isinstance(custom_id, str) or custom_id is None):
```

File: discord/ui/modal.py [hunk 3]
```diff
@@ -25,6 +25,10 @@ class Modal:
     """
 
     def __init__(self, title: str, custom_id: Optional[str] = None) -> None:
+
+        if not isinstance(custom_id, Union[str, None]):
```
Comment (Dorukyum) [inline discord/ui/modal.py:29]:
```suggestion
        if not (isinstance(custom_id, str) or custom_id is None):
```

File: discord/ui/select.py [hunk 4]
```diff
@@ -106,6 +106,10 @@ def __init__(
     ) -> None:
         super().__init__()
         self._selected_values: List[str] = []
+
+        if not isinstance(custom_id, Union[str, None]):
```
Comment (Dorukyum) [inline discord/ui/select.py:110]:
```suggestion
        if not (isinstance(custom_id, str) or custom_id is None):
```

**Review by Dorukyum — DISMISSED** [review]

File: discord/ui/input_text.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 from __future__ import annotations
 
 import os
-from typing import TYPE_CHECKING, Optional
+from typing import TYPE_CHECKING, Optional, Union
```
Comment (BobDotCom) [inline discord/ui/input_text.py:4]:
```suggestion
from typing import TYPE_CHECKING, Optional
```

**Review by BobDotCom — DISMISSED** [review]

File: discord/ui/input_text.py [hunk 6]
```diff
@@ -58,6 +58,10 @@ def __init__(
     ):
         super().__init__()
         custom_id = os.urandom(16).hex() if custom_id is MISSING else custom_id
+
+        if not (isinstance(custom_id, str) or custom_id is None):
+            raise TypeError(f"expected custom_id to be str, not {type(custom_id).__name__}")
```
Comment (UP929312) [inline discord/ui/input_text.py:63]: `type(custom_id).__name__}`
should be
`custom_id.__class__.__name__`

File: discord/ui/modal.py [hunk 7]
```diff
@@ -25,6 +25,10 @@ class Modal:
     """
 
     def __init__(self, title: str, custom_id: Optional[str] = None) -> None:
+
+        if not (isinstance(custom_id, str) or custom_id is None):
+            raise TypeError(f"expected custom_id to be str, not {type(custom_id).__name__}")
```
Comment (UP929312) [inline discord/ui/modal.py:30]: Same with this one,
`type(custom_id).__name__}`
should be
`custom_id.__class__.__name__`

File: discord/ui/select.py [hunk 8]
```diff
@@ -106,6 +106,10 @@ def __init__(
     ) -> None:
         super().__init__()
         self._selected_values: List[str] = []
+
+        if not (isinstance(custom_id, str) or custom_id is None):
+            raise TypeError(f"expected custom_id to be str, not {type(custom_id).__name__}")
```
Comment (UP929312) [inline discord/ui/select.py:111]: Same with this one,
`type(custom_id).__name__}`
should be
`custom_id.__class__.__name__`

Comment (krittick) [reply discord/ui/select.py:111]: No, this would cause the following result:

`TypeError: expected custom_id to be str, not type`

Comment (krittick) [reply discord/ui/select.py:111]: Nevermind, I misread. 

Regardless, either option has the same result. What's the benefit of your approach?

Comment (UP929312) [reply discord/ui/select.py:111]: Haha, no worries. I just think it's better form, seems to be more conventional to do it that way (even though they're essentially the same), I'd also say that type has to do a global lookup, while class is much better efficiency-wise (believe it's O(1) from a hashmap?), either way, it should be more efficient/faster and is more conventional

Comment (krittick) [reply discord/ui/select.py]: Good points, I've applied the changes. Thanks!

Comment (UP929312) [reply discord/ui/select.py]: No problem

File: discord/ui/input_text.py [hunk 9]
```diff
@@ -58,6 +58,10 @@ def __init__(
     ):
         super().__init__()
         custom_id = os.urandom(16).hex() if custom_id is MISSING else custom_id
+
```
Comment (Middledot) [inline discord/ui/input_text.py:61]:
```suggestion
```

File: discord/ui/modal.py [hunk 10]
```diff
@@ -26,6 +26,10 @@ class Modal:
     """
 
     def __init__(self, title: str, custom_id: Optional[str] = None) -> None:
+
```
Comment (Middledot) [inline discord/ui/modal.py:29]:
```suggestion
```

**Review by Middledot — COMMENTED** [review]
extra whitespaces

**Review by Middledot — APPROVED** [review]

**Review by Dorukyum — APPROVED** [review]

**Pull request merged**